### DISCUSSION TITLE
Optional mutation type for the parameter of `select` in `MutationStateOptions`

### DISCUSSION THREAD

**Discussion by mohamedhamza12** [TanStack/query#8047] [Ideas]
Hey, I'm currently working with gRPC endpoints using grpc-web and @tanstack/react-query. Because of this, if a mutation fails the type of error it returns is `RpcError`. I'm using `useMutationState` to return a an array of mutations but passing `select` to its options to return the error of each mutation always expects the error to be of type `Error`. My suggestion is passing an extra optional generic type parameter to `useMutationState` representing the mutation type, so something similar to the following:

```ts
type MutationStateOptions<TResult = MutationState, TMutation = Mutation> = {
    filters?: MutationFilters;
    select?: (mutation: TMutation) => TResult;
};
declare function useMutationState<TResult = MutationState, TMutation = Mutation>(options?: MutationStateOptions<TResult, TMutation>, queryClient?: QueryClient): Array<TResult>;
```

This would enable me to use the hook as follows:

```ts
  const mutationState = useMutationState<RpcError | null, Mutation<unknown, RpcError>>({
    select(mutation) {
      return mutation.state.error;
    },
  });
```

**Comment by TkDodo** [comment]
if you want to narrow error states, it has to be either [globally with module augmentation](https://tanstack.com/query/v5/docs/framework/react/typescript#registering-a-global-error) or by doing runtime type narrowing (which is preferred)

**Reply by mohamedhamza12** [reply]
Well yes, that makes sense in my specific use case. More generally though, wouldn't it be good to be able to specify what the type of each mutation is specifically? For example, if you would want to change the type of other generic type parameters, not just the error type particularly
